### PR TITLE
pci: allow for dumping and setting of config registers

### PR DESCRIPTION
**Pull request by rminnich**
Back in the day, this pci command would have been incredibly handy.

Suppose you want to dump 16 bits @ 0 and 2, and 32 bits at 4, and 8 bits at 4.
This is device and vendor id, and 32 bits @ 4, and 8 bits of command register.
And you want it for all devices at device 18, a root port.

pci -s *18* 0.w 2.w 4.l 4.b

0000:00:18.0: VMware PCI Express Root Port
00000000:15ad
00000002:07a0
00000004:00100007
00000004:07
0000:00:18.1: VMware PCI Express Root Port
00000000:15ad
00000002:07a0
00000004:00100007
00000004:07
0000:00:18.2: VMware PCI Express Root Port
00000000:15ad
00000002:07a0
00000004:00100007
00000004:07
0000:00:18.3: VMware PCI Express Root Port
00000000:15ad
00000002:07a0
00000004:00100007
00000004:07
0000:00:18.4: VMware PCI Express Root Port
00000000:15ad
00000002:07a0
00000004:00100007
00000004:07
0000:00:18.5: VMware PCI Express Root Port
00000000:15ad
00000002:07a0
00000004:00100007
00000004:07
0000:00:18.6: VMware PCI Express Root Port
00000000:15ad
00000002:07a0
00000004:00100007
00000004:07
0000:00:18.7: VMware PCI Express Root Port
00000000:15ad
00000002:07a0
00000004:00100007
00000004:07

Here is how you can set, and then see what value was really set:
 sudo ./pci -s 0000:00:18.7 4.b=7 4.b
0000:00:18.7: VMware PCI Express Root Port
00000004:07

The *18* version of this command, shown above, would also work here. I've been using bash scripts
looping over TBFN pci notation for 15 years; no longer.

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

File: cmds/pci/pci.go [hunk 1]
```diff
@@ -23,8 +27,84 @@ var (
...
+			continue
+		}
+		s := 32
+		if len(rs) == 2 {
+			switch rs[1] {
+			default:
+				log.Printf("Bad size: %v. Due to this error no more commands will be issued", rs[1])
+				justCheck = true
+				continue
+			case "l":
```
Comment (hugelgupf) [inline cmds/pci/pci.go]: case "l":
  fallthrough

otherwise l does nothing?

File: cmds/pci/pci.go [hunk 2]
```diff
@@ -23,8 +27,84 @@ var (
...
+			case "w":
+				s = 16
+			case "b":
+				s = 8
+			}
+		}
+		if justCheck {
```
Comment (hugelgupf) [inline cmds/pci/pci.go]: I'm not sure I understand the justCheck... it's set to true below, but checked up here?

File: pkg/pci/devices.go [hunk 3]
```diff
@@ -38,3 +38,26 @@ func (d Devices) ReadConfig() error {
 	}
 	return nil
 }
+
+// ReadConfigRegister reads the config info for all the devices.
```
Comment (hugelgupf) [inline pkg/pci/devices.go]: what are r and s? call them offset and size?

File: pkg/pci/devices.go [hunk 4]
```diff
@@ -38,3 +38,26 @@ func (d Devices) ReadConfig() error {
...
+		if err != nil {
+			return nil, err
+		}
+		vals = append(vals, v)
+	}
+	return vals, nil
+}
+
+// WriteConfigRegister writes the config info for all the devices.
+func (d Devices) WriteConfigRegister(r, s int64, val uint64) error {
```
Comment (hugelgupf) [inline pkg/pci/devices.go]: offset, size, val?

File: pkg/pci/pci.go [hunk 5]
```diff
@@ -46,3 +48,74 @@ func (p *PCI) ReadConfig() error {
...
+	f, err := os.Open(filepath.Join(p.FullPath, "config"))
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	var reg uint64
+	r := &barreg{offset: offset, File: f}
+	switch size {
+	default:
+		return 0, fmt.Errorf("%d is not valid: only options are 8, 16, 32, 64")
```
Comment (bkochendorfer) [inline pkg/pci/pci.go]: `pkg/pci/pci.go:77: missing argument for Errorf("%d"): format reads arg 1, have only 0 args`

File: pkg/pci/pci.go [hunk 6]
```diff
@@ -46,3 +48,74 @@ func (p *PCI) ReadConfig() error {
...
+func (p *PCI) WriteConfigRegister(offset, size int64, val uint64) error {
+	f, err := os.OpenFile(filepath.Join(p.FullPath, "config"), os.O_WRONLY, 0)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	w := &barreg{offset: offset, File: f}
+	switch size {
+	default:
+		return fmt.Errorf("%d is not valid: only options are 8, 16, 32, 64")
```
Comment (bkochendorfer) [inline pkg/pci/pci.go]: `pkg/pci/pci.go:107: missing argument for Errorf("%d"): format reads arg 1, have only 0 args`

File: cmds/pci/pci.go (hunk 1 above)
Comment (rminnich) [reply cmds/pci/pci.go]: This is intended. s is set to 32 above. There is no need to do anything else, but I left this placeholder.

File: cmds/pci/pci.go (hunk 2 above)
Comment (rminnich) [reply cmds/pci/pci.go]: you can go through this loop once per command. justCheck is set for any errors and subsequent commands are ignored.

File: pkg/pci/devices.go (hunk 3 above)
Comment (rminnich) [reply pkg/pci/devices.go]: reg, size. Done.

File: pkg/pci/devices.go (hunk 4 above)
Comment (rminnich) [reply pkg/pci/devices.go]: done

File: pkg/pci/pci.go (hunk 5 above)
Comment (rminnich) [reply pkg/pci/pci.go]: done

File: pkg/pci/pci.go (hunk 6 above)
Comment (rminnich) [reply pkg/pci/pci.go]: done

**Comment by hugelgupf** [comment]
fwiw i find the command-line syntax a little unintuitive or hard to remember, but not sure what else to do about it.

**Comment by rminnich** [comment]
If you're used to setpci, lspci, etc. it's actually modeled on that ... once you're used to it it's quite convenient.

**Pull request merged**